### PR TITLE
Check whether index is valid at runtime

### PR DESCRIPTION
**Pull request by yixinglu**
as title.

### CODE REVIEW

**Review by jievince — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

File: src/executor/query/IndexScanExecutor.cpp [hunk 1]
```diff
@@ -28,8 +30,16 @@ folly::Future<Status> IndexScanExecutor::indexScan() {
...
         return finish(ResultBuilder().value(Value(std::move(dataSet))).finish());
     }
+
+    const auto &ictxs = lookup->queryContext();
+    auto iter = std::find_if(
+        ictxs.begin(), ictxs.end(), [](auto &ictx) { return !ictx.index_id_ref().is_set(); });
+    if (ictxs.empty() || iter != ictxs.end()) {
+        return Status::Error("There is no index to use at runtime");
+    }
+
```
Comment (bright-starry-sky) [inline src/executor/query/IndexScanExecutor.cpp:40]: Why is the index ID not set?

**Comment by bright-starry-sky** [comment]
In addition, maybe need to check the validity of the index. for example :
tag t1 (c1)
index i1 ()
this index is only used for count(*)

File: src/executor/query/IndexScanExecutor.cpp [hunk 2]
```diff
@@ -28,8 +30,16 @@ folly::Future<Status> IndexScanExecutor::indexScan() {
         DataSet dataSet({"dummy"});
         return finish(ResultBuilder().value(Value(std::move(dataSet))).finish());
     }
+
+    const auto &ictxs = lookup->queryContext();
+    auto iter = std::find_if(
+        ictxs.begin(), ictxs.end(), [](auto &ictx) { return !ictx.index_id_ref().is_set(); });
+    if (ictxs.empty() || iter != ictxs.end()) {
```
Comment (bright-starry-sky) [inline src/executor/query/IndexScanExecutor.cpp:37]: Is the checking logic in the if statement` ictxs.empty()` necessary?

File: src/executor/query/IndexScanExecutor.cpp (hunk 1 above)
Comment (yixinglu) [reply src/executor/query/IndexScanExecutor.cpp:40]: Now index is selected in multiple opt rules. when all rules could not select the index, IndexFullScanNode without index is still in the execution plan. so here needs to check the real index at runtime.

File: src/executor/query/IndexScanExecutor.cpp (hunk 2 above)
Comment (yixinglu) [reply src/executor/query/IndexScanExecutor.cpp:37]: My test cases is for this scenario.

**Comment by yixinglu** [comment]
> In addition, maybe need to check the validity of the index. for example :
> tag t1 (c1)
> index i1 ()
> this index is only used for count(*)

Now if you use `t1.c1 to` lookup, the execution plan can use the index i1 to full scan the whole tag. such as
```
(root@nebula) [test_index]> explain lookup on player where player.name == 'tim'
Execution succeeded (time spent 1403/1847 us)

Execution Plan (optimize time 813 us)

-----+------------------+--------------+----------------+--------------------------------------
| id | name             | dependencies | profiling data | operator info                       |
-----+------------------+--------------+----------------+--------------------------------------
|  3 | Project          | 2            |                | outputVar: [                        |
|    |                  |              |                |   {                                 |
|    |                  |              |                |     "colNames": [                   |
|    |                  |              |                |       "VertexID"                    |
|    |                  |              |                |     ],                              |
|    |                  |              |                |     "name": "__Project_3",          |
|    |                  |              |                |     "type": "DATASET"               |
|    |                  |              |                |   }                                 |
|    |                  |              |                | ]                                   |
|    |                  |              |                | inputVar: __Filter_2                |
|    |                  |              |                | columns: [                          |
|    |                  |              |                |   "$-.VertexID AS VertexID"         |
|    |                  |              |                | ]                                   |
-----+------------------+--------------+----------------+--------------------------------------
|  2 | Filter           | 4            |                | outputVar: [                        |
|    |                  |              |                |   {                                 |
|    |                  |              |                |     "colNames": [                   |
|    |                  |              |                |       "VertexID",                   |
|    |                  |              |                |       "player.name"                 |
|    |                  |              |                |     ],                              |
|    |                  |              |                |     "type": "DATASET",              |
|    |                  |              |                |     "name": "__Filter_2"            |
|    |                  |              |                |   }                                 |
|    |                  |              |                | ]                                   |
|    |                  |              |                | inputVar: __TagIndexFullScan_1      |
|    |                  |              |                | condition: (player.name=="tim")     |
|    |                  |              |                | isStable: false                     |
-----+------------------+--------------+----------------+--------------------------------------
|  4 | TagIndexFullScan | 0            |                | outputVar: [                        |
|    |                  |              |                |   {                                 |
|    |                  |              |                |     "colNames": [                   |
|    |                  |              |                |       "VertexID",                   |
|    |                  |              |                |       "player.name"                 |
|    |                  |              |                |     ],                              |
|    |                  |              |                |     "name": "__TagIndexFullScan_1", |
|    |                  |              |                |     "type": "DATASET"               |
|    |                  |              |                |   }                                 |
|    |                  |              |                | ]                                   |
|    |                  |              |                | inputVar:                           |
|    |                  |              |                | space: 5600                         |
|    |                  |              |                | dedup: false                        |
|    |                  |              |                | limit: 9223372036854775807          |
|    |                  |              |                | filter:                             |
|    |                  |              |                | orderBy: []                         |
|    |                  |              |                | schemaId: 5601                      |
|    |                  |              |                | isEdge: false                       |
|    |                  |              |                | returnCols: [                       |
|    |                  |              |                |   "_vid",                           |
|    |                  |              |                |   "name"                            |
|    |                  |              |                | ]                                   |
|    |                  |              |                | indexCtx: [                         |
|    |                  |              |                |   {                                 |
|    |                  |              |                |     "columnHints": [],              |
|    |                  |              |                |     "index_id": 5602,               |
|    |                  |              |                |     "filter": ""                    |
|    |                  |              |                |   }                                 |
|    |                  |              |                | ]                                   |
-----+------------------+--------------+----------------+--------------------------------------
|  0 | Start            |              |                | outputVar: [                        |
|    |                  |              |                |   {                                 |
|    |                  |              |                |     "colNames": [],                 |
|    |                  |              |                |     "type": "DATASET",              |
|    |                  |              |                |     "name": "__Start_0"             |
|    |                  |              |                |   }                                 |
|    |                  |              |                | ]                                   |
-----+------------------+--------------+----------------+--------------------------------------
```

**Pull request merged**